### PR TITLE
Allow rescorer with field collapsing

### LINKED ISSUE #27243: support field collapsing + rescore
I have some searches that require field collapsing. I need the feature, and I also need the performance improvement over using parent/child or nested. I would like to start using the [learning to rank plugin](https://github.com/o19s/elasticsearch-learning-to-rank) for these searches but rescore support is needed to have good performance, at least for my use case. Any chance of adding this?

### PR DESCRIPTION
**Pull request by jimczi**
This change adds the support for rescoring collapsed documents. 
The rescoring is applied on the top document per group on each shard.

Closes #27243

### CODE REVIEW

**Review by mark-vieira — APPROVED** [review]

File: rest-api-spec/build.gradle [hunk 1]
```diff
@@ -83,6 +83,7 @@ tasks.named("yamlRestTestV7CompatTransform").configure { task ->
   task.skipTest("search/370_profile/fetch source", "profile output has changed")
   task.skipTest("search/370_profile/fetch nested source", "profile output has changed")
   task.skipTest("search/240_date_nanos/doc value fields are working as expected across date and date_nanos fields", "Fetching docvalues field multiple times is no longer allowed")
+  task.skipTest("search/110_field_collapsing/field collapsing and rescore", "#107779 Field collapsing is compatible with rescore in 8.15")
```
Comment (benwtrent) [inline rest-api-spec/build.gradle:86]: Do you still need to skip when the test is deleted?

File: server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java [hunk 2]
```diff
@@ -24,28 +29,65 @@ public class RescorePhase {
...
+                topDocs = rewriteTopGroups(topGroups, topDocs);
+            }
             context.queryResult()
                 .topDocs(new TopDocsAndMaxScore(topDocs, topDocs.scoreDocs[0].score), context.queryResult().sortValueFormats());
         } catch (IOException e) {
             throw new ElasticsearchException("Rescore Phase Failed", e);
         }
     }
 
+    private static TopFieldGroups rewriteTopGroups(TopFieldGroups originalTopGroups, TopDocs rescoredTopDocs) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java:68]: Ah, this is effectively re-ordering the top field groups based on the rescored doc order?

File: server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java [hunk 3]
```diff
@@ -24,28 +29,65 @@ public class RescorePhase {
...
+        assert originalTopGroups.fields.length == 1 && SortField.FIELD_SCORE.equals(originalTopGroups.fields[0])
+            : "rescore must always sort by score descending";
+        Map<Integer, Object> docIdToGroupValue = new HashMap<>();
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java:71]: quibble, maybe `Maps.newMapWithExpectedSize`?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/110_field_collapsing.yml [hunk 4]
```diff
@@ -281,24 +281,6 @@ setup:
   - match: { hits.hits.1.fields.numeric_group: [1] }
   - match: { hits.hits.1.sort: [1] }
 
----
```
Comment (benwtrent) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/110_field_collapsing.yml:284]: Could we have a positive test case in yaml? This gives us way more coverage than IT tests.

**Review by benwtrent — COMMENTED** [review]
Very exciting!

File: server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java:68]: Right, the rescoring works with `TopDocs` so we need to reconstruct the `TopFieldGroups`.

File: server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java (hunk 3 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java]: Addressed in https://github.com/elastic/elasticsearch/pull/107779/commits/c1fb5e33186a7142bd588e1345004538fc2ab98c

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/110_field_collapsing.yml (hunk 4 above)
Comment (jimczi) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/110_field_collapsing.yml:284]: Added in https://github.com/elastic/elasticsearch/pull/107779/commits/c1fb5e33186a7142bd588e1345004538fc2ab98c

File: rest-api-spec/build.gradle (hunk 1 above)
Comment (jimczi) [reply rest-api-spec/build.gradle]: It seems so yes, these tests are sourced from the previous version afaik.

**Review by benwtrent — APPROVED** [review]
Fantastic.

Do we think we could also add rescore support for inner hits in the future?

I realize inner-hits are optional for collapse, so we can discuss this later. But it would be good to provide rescoring for inner hits so that their scores & order reflect their collapsed fields. This may require an API change for the inner-hits configuration to accept a "rescore" parameter or something. I honestly don't know.

**Comment by jimczi** [comment]
> Do we think we could also add rescore support for inner hits in the future

That's a good question. Adding the rescoring for `inner_hits` should be trivial since we can copy the rescorer into the expand query. I wanted to add that initially but then realise that users may want to apply a different window_size when rescoring inside each group. The original window_size might be too big to apply on every group since that would be applied on each top docs request per group. I left it for a follow up and added a warning in the doc for now.

File: docs/reference/search/search-your-data/collapse-search-results.asciidoc [hunk 5]
```diff
@@ -175,6 +174,65 @@ GET /my-index-000001/_search
 ----
 // TEST[setup:my_index]
 
+[discrete]
+[[rescore-collapse-results]]
+==== Rescore collapse results
+
+You can use field collapsing alongside the <<rescore, `rescore`>> search parameter.
+Rescorers runs on every shard for the top-ranked document per collapsed field.
```
Comment (kderusso) [inline docs/reference/search/search-your-data/collapse-search-results.asciidoc:182]:
```suggestion
Rescorers run on every shard for the top-ranked document per collapsed field.
```

**Review by kderusso — APPROVED** [review]
LGTM - thanks for taking care of this!

**Pull request merged**